pkg/e2e/infra/infra.go

```diff
@@ -446,4 +476,53 @@ var _ = g.Describe("[Feature:Machines] Managed cluster should", func() {
 		o.Expect(err).NotTo(o.HaveOccurred())
 
 	})
+
+	g.It("reject invalid machinesets", func() {
+		var err error
+		client, err := e2e.LoadClient()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("Creating invalid machineset")
+		invalidMachineSet := invalidMachinesetWithEmptyProviderConfig()
+
+		err = client.Create(context.TODO(), invalidMachineSet)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("Waiting for ReconcileError MachineSet event")
+		err = wait.PollImmediate(e2e.RetryMedium, e2e.WaitShort, func() (bool, error) {
+			eventList := corev1.EventList{}
+			if err := client.List(context.TODO(), nil, &eventList); err != nil {
+				glog.Errorf("error querying api for eventList object: %v, retrying...", err)
+				return false, nil
+			}
+
+			glog.Infof("Fetching ReconcileError MachineSet invalid-machineset event")
+			for _, event := range eventList.Items {
+				if event.Reason != "ReconcileError" || event.InvolvedObject.Kind != "MachineSet" || event.InvolvedObject.Name != invalidMachineSet.Name {
+					continue
+				}
+
+				glog.Infof("Found ReconcileError event for %q machine set with the following message: %v", event.InvolvedObject.Name, event.Message)
+				return true, nil
+			}
+
+			return false, nil
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// Verify the number of machines does not grow over time.
+		// The assumption is once the ReconcileError event is recorded and caught,
+		// the machineset is not reconciled again until it's updated.
+		machineList := &mapiv1beta1.MachineList{}
+		err = client.List(context.TODO(), runtimeclient.MatchingLabels(invalidMachineSet.Spec.Template.Labels), machineList)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By(fmt.Sprintf("Verify no machine from %q machineset were created", invalidMachineSet.Name))
+		glog.Infof("Have %v machines generated from %q machineset", len(machineList.Items), invalidMachineSet.Name)
+		o.Expect(len(machineList.Items)).To(o.BeNumerically("==", 0))
```

Comment:
We might wait here for 10 seconds just to make sure no machine is created even after machineset controller recorded `ReconcileError` event.